Adds optimize CLS article (#2715)
GoogleChrome/web.dev: src/site/content/en/blog/optimize-cls/index.md

---
title: Optimize Cumulative Layout Shift
subhead: Learn how to avoid sudden layout shifts to improve user-experience
authors:
  - addyosmani
date: 2020-05-05

> hero: hero.png

Author: Thank you for the suggestion, @kaycebasques. If I understood correctly, the change here was to drop the circles as it made it look like the user had already tapped.

![image](https://user-images.githubusercontent.com/110953/80743154-58277400-8ad1-11ea-966e-341c8cdf227f.png)

One thing I was unsure of was if you just wanted them dropped for the first and third images or for all of them (so latest commit removes the circles for all). If you feel they should remain for the second and forth images, let me know :)

> hero: hero.png

Reviewer: Sorry, words are hard. I meant that the circles should remain on the second and fourth images from the left. Those circles convey the idea that the user has tapped the screen.

> hero: hero.png

Author: Ack. Next commits will address this :)

description: |
  Cumulative Layout Shift (CLS) is a metric that quantifies how often users experience sudden shifts in page content. In this guide, we'll cover optimizing common causes of CLS such as images and iframes without dimensions or dynamic content.
alt: Layout shifts can suddenly push the content you are reading or are about to click further down the page, leading to a poor user-experience. Reserving space for dynamic content causing layout shifts leads to a more delightful user experience.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - fast

>   - web vitals
>   - cumulative layout shift

Reviewer: ```suggestion
  - web-vitals
  - cumulative-layout-shift
  - cls
```

```html
<!-- set a 640:360 i.e a 16:9 - aspect ratio -->
<img src="puppy.jpg" width="640" height="360" alt="Puppy with balloons">
```

…and the [UA stylesheets](https://developer.mozilla.org/docs/Web/CSS/Cascade#User-agent_stylesheets) of all browsers add a [default aspect ratio](https://html.spec.whatwg.org/multipage/rendering.html#attributes-for-embedded-content-and-images) based on the element's existing `width` and `height` attributes:

```css
img {
    aspect-ratio: attr(width) / attr(height);
}
```

This calculates an aspect ratio based on the `width` and `height` attributes before the image has loaded. It provides this information at the very start of layout calculation. As soon as an image is told to be a certain width (for example `width: 100%`), the aspect ratio is used to calculate the height.

Tip: If you're having a hard time understanding aspect ratio, a handy [calculator](https://aspectratiocalculator.com/16-9.html) is available to help.

> The above image aspect ratio changes have shipped in [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1547231), [Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=979891) and are coming to [WebKit](https://twitter.com/smfr/status/1220051332767174656) (Safari).

Reviewer: ```suggestion
The above image aspect ratio changes have shipped in [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1547231) and [Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=979891), and are coming to [WebKit](https://twitter.com/smfr/status/1220051332767174656) (Safari).
```
Small grammar fix

```html
<picture>
  <source media="(max-width: 799px)" srcset="puppy-480w-cropped.jpg">
  <source media="(min-width: 800px)" srcset="puppy-800w.jpg">
  <img src="puppy-800w.jpg" alt="Puppy with balloons">
</picture>
```

It's very possible these images could have different aspect ratios and browsers are still evaluating what the most efficient solution here should be, including if dimensions should be specified on all sources. Until a solution is decided on, relayout is still possible here.

## Ads, embeds and iframes without dimensions 📢😱

### Advertisements

Ads are one of the largest contributors to layout shifts on the web. Ad networks and publishers often support dynamic ad sizes. Ad sizes increase performance/revenue due to higher click rates and more ads competing in the auction. Unfortunately, this can lead to a suboptimal user experience due to ads pushing content you're reading down the page.

During the ad lifecycle, many points can introduce layout shift:

*   When a site inserts the ad container in the DOM
*   When a site resizes the ad container with first-party code
*   When the ad tag library loads (and resizes the ad container)

> *   When the ad fills a container (resizes if the final ad has a different size)

Reviewer: ```suggestion
*   When the ad fills a container (and resizes if the final ad has a different size)
```
@addyosmani I think there's supposed to an `and` here but not sure

The good news is that it's possible for sites to follow best practices to reduce ad shift. Sites can mitigate these layout shifts by:

> * Statically reserving space for the ad slot
>   * In other words, style the element before the ad tag library loads
>   * If placing ads in the content flow, ensure shifts are eliminated by reserving the slot size. These ads _shouldn't_ cause layout shifts if loaded off-screen.
> * Take care when placing non-sticky ads near the top of the viewport. 
>   * In the below example, it's recommended to move the ad to below the "world vision" logo and make sure to reserve enough space for the slot.
> * Eliminate shifts by reserving the largest possible size for the ad slot
>   * Works, but risks having a blank space if a smaller ad creative fills the slot
> * Choose the most likely size for the ad slot based on historical data

Reviewer: ```suggestion
* Statically reserve space for the ad slot.
  * In other words, style the element before the ad tag library loads.
  * If placing ads in the content flow, ensure shifts are eliminated by reserving the slot size. These ads _shouldn't_ cause layout shifts if loaded off-screen.
* Take care when placing non-sticky ads near the top of the viewport. 
  * In the below example, it's recommended to move the ad to below the "world vision" logo and make sure to reserve enough space for the slot.
* Eliminate shifts by reserving the largest possible size for the ad slot.
  * This works, but it risks having a blank space if a smaller ad creative fills the slot.
* Choose the most likely size for the ad slot based on historical data.
```

Some sites may find collapsing the slot initially can reduce layout shifts if the ad slot is unlikely to fill. There isn't an easy way to choose the exact size each time, unless you control the ad serving yourself.

  <figure class="w-figure">
    <video controls loop muted class="w-screenshot" poster="video-final-ads-no-dimensions.jpg">
      <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/final_ads_no_dimensions.webm" type="video/webm;">
      <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/final_ads_no_dimensions.mp4" type="video/mp4; codecs=h264">
    </video>
   <figcaption class="w-figcaption">
      Ads without sufficient space reserved.
    </figcaption>
  </figure>

  <figure class="w-figure">

>     <video controls loop muted class="w-screenshot" poster="video-top-ad-after.jpg">
>       <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/top-ad-after.mp4" type="video/mp4; codecs=h264">
>     </video>

Reviewer: ```suggestion
    <video controls loop muted class="w-screenshot" poster="video-top-ad-after.jpg">
      <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/top-ad-after.mp4" type="video/mp4; codecs=h264">
      <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/top-ad-after.webm" type="video/webm;">
    </video>
```

   <figcaption class="w-figcaption">
      Ads with sufficient space reserved.
    </figcaption>
  </figure>

<figure class="w-figure">
  <img src="Optimize-Cumulative6.png " alt="Lighthouse report showing the before/after impact to Cumulative Layout Shift of reserving space for banners like ads">
  <figcaption class="w-figcaption">
    Lighthouse 6.0 impact of reserving space for this banner on CLS
  </figcaption>
</figure>

#### Statically reserve space for the ad slot

Statically style slot DOM elements with the same sizes passed to your tag library. This can help ensure the library doesn't introduce layout shifts when it loads. If you don't do this, the library may change the size of the slot element after page layout.

> Also consider the sizes of smaller ad serves. If a smaller ad is served, a publisher can style the (larger) container to avoid layout shifts. The downside to this approach will increase the amount of blank space, so keep in mind the trade-off here.

Reviewer: ```suggestion
Also consider the sizes of smaller ad serves. If a smaller ad is served, a publisher can style the (larger) container to avoid layout shifts. The downside to this approach is that it will increase the amount of blank space, so keep in mind the trade-off here.
```

### Developer Tools 🔧

I'm happy to share there are a number of tools available to measure and debug Cumulative Layout Shift (CLS).

[Lighthouse](https://developers.google.com/web/tools/lighthouse) [6.0](https://github.com/GoogleChrome/lighthouse/releases) and above include support for measuring CLS in a lab setting. This release will also highlight the nodes that cause the most layout shifting.

![alt_text](Optimize-Cumulative11.jpg "image_tooltip")

The [Performance panel](https://developers.google.com/web/tools/chrome-devtools/evaluate-performance) in DevTools highlights layout shifts in the **Experience** section as of Chrome 84. The **Summary** view for a `Layout Shift` record includes the cumulative layout shift score as well as a rectangle overlay showing the affected regions.

<figure class="w-figure">
  <img src="Optimize-Cumulative12.jpg" alt="Layout Shift records being displayed in the Chrome DevTools performance panel when expanding the Experience section">
  <figcaption class="w-figcaption">

> After recording a new trace in the Performance panel, the 'Experience' section of the results is populated with a red-tinted bar displaying a "Layout Shift" record. Clicking the record allows you to drill down into impacted elements (e.g note the moved from/to entries)

Reviewer: ```suggestion
After recording a new trace in the Performance panel, the <b>Experience</b> section of the results is populated with a red-tinted bar displaying a <code>Layout Shift</code> record. Clicking the record allows you to drill down into impacted elements (e.g. note the moved from/to entries).
```

date: 2020-05-05
hero: hero.png
description: |
  Cumulative Layout Shift (CLS) is a metric that quantifies how often users experience sudden shifts in page content. In this guide, we'll cover optimizing common causes of CLS such as images and iframes without dimensions or dynamic content.
alt: Layout shifts can suddenly push the content you are reading or are about to click further down the page, leading to a poor user-experience. Reserving space for dynamic content causing layout shifts leads to a more delightful user experience.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - fast
  - web-vitals
  - cumulative-layout-shift
  - cls
  - lighthouse
---

"I was about to click that! Why did it move? 😭"

Layout shifts can be distracting to users. Imagine you've started reading an article when all of a sudden elements shift around the page, throwing you off and requiring you to find your place again. This is very common on the web, including when reading the news, or trying to click those 'Search' or 'Add to Cart' buttons. Such layout shifts happen when elements on a page move because another element was suddenly added to the page or resized. 

> [Cumulative Layout Shift](/cls) (CLS) - a [Core Web Vitals](/metrics) metric, measures the instability of content by summing shift scores across layout shifts that don't occur within 500ms of user input. It looks at how much the viewport shifted as well as the distance the element impacted was shifted.

Reviewer 2: Also, can you link to `/vitals/`, which should be live before you merge this PR?

Author: Certainly. Updating now.

In this guide, we'll cover optimizing common causes of layout shifts.

> <img src="cls-graphic.svg" alt="Cumulative layout shift measures layout shifts to ensure users experience smooth and natural interactions. In this graphic, we highlight sites should strive to have a CLS of less than 0.1 while a CLS of over 0.25 is considered poor.">

Reviewer 2: Do you think it makes sense to use the same markup and image I have here? This should be fine as long as my PR is merged first: https://github.com/GoogleChrome/web.dev/pull/2764/files#diff-0bf26a70d5c42491995a4aaa6f7cfa23R78-R84

Reviewer 2: cc: @housseindjirdeh

Reviewer 3: Made the change here: https://github.com/GoogleChrome/web.dev/pull/2793/files#diff-df5d0e12cb0f29cf54743f960761dfb9R25-R31

@addyosmani Let me know if you would rather not however!

**Summary:** Always include `width` and `height` size attributes on your images and video elements. Alternatively, reserve the required space with [CSS aspect ratio boxes](https://css-tricks.com/aspect-ratio-boxes/). This approach ensures that the browser can allocate the correct amount of space in the document while the image is loading.

  <figure class="w-figure">

>     <video controls loop muted class="w-screenshot" poster="video-image-dimensions.jpg">

Author: Opted for a poster + manual playback here to avoid readers downloading 30MB+ of MP4s. Interested in a confirm we're okay with this.

Reviewer 4: that sounds ok to me

  <figure class="w-figure">
    <video controls loop muted class="w-screenshot" poster="video-image-dimensions.jpg">

>       <source src="https://storage.googleapis.com/web-dev-assets/optimize-cls/image-dimensions.webm" type="video/webm;">

Author: WebM videos here won't load until they've been uploaded to the Cloud bucket. Pinged Kayce to request :)

Reviewer: Done
